core/src/com/unciv/logic/multiplayer/chat/ChatWebSocketManager.kt

```diff
@@ -0,0 +1,249 @@
+        ) {
+            relayGlobalMessage("Authentication issue detected! You have to set a password to use Chat.")
+        }
+
+        restartSocket(true)
+    }
+
+    suspend fun startSession() {
+        try {
+            session?.close()
+            session = client.webSocketSession {
+                url(chatUrl)
+                header(HttpHeaders.Authorization, UncivGame.Current.settings.multiplayer.getAuthHeader())
+            }
+
+            session!!.runCatching {
+                if (isActive) {
+                    if (errorReconnectionAttempts == 0) {
+                        println("ChatLog: Connected to WebSocket.")
+                        relayGlobalMessage("Successfully connected to WebSocket server!")
+                    } else if (errorReconnectionAttempts > 0) {
+                        println("ChatLog: Re-established webSocket connection.")
+                        relayGlobalMessage("Successfully re-established WebSocket connection!")
+                    }
+                }
+
+                val gameIds = ChatStore.getGameIds()
+                    .union(UncivGame.Current.onlineMultiplayer.games.mapNotNull { it.preview?.gameId })
+                this.sendSerialized(Message.Join(gameIds.toList()))
+
+                while (this.isActive) {
+                    val response = receiveDeserialized<Response>()
+                    when (response) {
+                        is Response.Chat -> EventBus.send(
+                            ChatMessageReceived(
+                                response.gameId, response.civName, response.message
+                            )
+                        )
+
+                        is Response.Error -> relayGlobalMessage("Error: ${response.message}", "Server")
+                        is Response.JoinSuccess -> Unit
+                    }
+                    yield()
+                }
+            }
+                .onSuccess { restartSocket() }
+                .onFailure { handleWebSocketThrowables(it) }
+        } catch (e: Exception) {
+            handleWebSocketThrowables(e)
+        }
+    }
+
+    private fun startSocket() {
+        if (job?.isActive == true) return
+        job?.cancel()
+        job = Concurrency.run("MultiplayerChat") { startSession() }
+    }
+
+    /**
+     * By default, this gets autocancelled if the [job] is still running.
+     * Force mode will cancel the previous [job] and reassign a new one.
+     * This is helpfull when we need to reset [job] due to events.
+     */
+    @OptIn(DelicateCoroutinesApi::class)
+    private fun restartSocket(dueToError: Boolean = false, force: Boolean = false) {
+        if (dueToError) {
+            if (errorReconnectionAttempts++ > MAX_ERROR_RECONNECTION_ATTEMPTS) {
+                return
+            }
+        } else errorReconnectionAttempts = 0
+
+        GlobalScope.launch {
+            flow {
+                delay(RECONNECT_TIME_MS)
+                emit(Unit)
```

Comment:
What is going on here?

Reply:
We are just trying to reconnect after a delay. This is the coroutines way to do that apparently.

Reply:
fixed in https://github.com/yairm210/Unciv/pull/13599/commits/1be3151be60853554e1f30e63b4051e03f82f875